coderd/prometheusmetrics/prometheusmetrics.go

```diff
@@ -283,8 +283,177 @@ func Agents(ctx context.Context, logger slog.Logger, registerer prometheus.Regis
+		Help:      "The number of established connections by agent",
+	}, []string{"agent_name", "username", "workspace_name"}))
+	err = registerer.Register(agentStatsConnectionCountGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	agentStatsConnectionMedianLatencyGauge := NewCachedGaugeVec(prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agentstats",
+		Name:      "connection_median_latency_seconds",
+		Help:      "The median agent connection latency in seconds",
+	}, []string{"agent_name", "username", "workspace_name"}))
+	err = registerer.Register(agentStatsConnectionMedianLatencyGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	agentStatsSessionCountJetBrainsGauge := NewCachedGaugeVec(prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agentstats",
+		Name:      "session_count_jetbrains",
+		Help:      "The number of session established by JetBrains",
+	}, []string{"agent_name", "username", "workspace_name"}))
+	err = registerer.Register(agentStatsSessionCountJetBrainsGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	agentStatsSessionCountReconnectingPTYGauge := NewCachedGaugeVec(prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agentstats",
+		Name:      "session_count_reconnecting_pty",
+		Help:      "The number of session established by reconnecting PTY",
+	}, []string{"agent_name", "username", "workspace_name"}))
+	err = registerer.Register(agentStatsSessionCountReconnectingPTYGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	agentStatsSessionCountSSHGauge := NewCachedGaugeVec(prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agentstats",
+		Name:      "session_count_ssh",
+		Help:      "The number of session established by SSH",
+	}, []string{"agent_name", "username", "workspace_name"}))
+	err = registerer.Register(agentStatsSessionCountSSHGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	agentStatsSessionCountVSCodeGauge := NewCachedGaugeVec(prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "coderd",
+		Subsystem: "agentstats",
+		Name:      "session_count_vscode",
+		Help:      "The number of session established by VSCode",
+	}, []string{"agent_name", "username", "workspace_name"}))
+	err = registerer.Register(agentStatsSessionCountVSCodeGauge)
+	if err != nil {
+		return nil, err
+	}
+
+	createdAfter := initialCreateAfter
+	ctx, cancelFunc := context.WithCancel(ctx)
+	ticker := time.NewTicker(duration)
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+			}
+
+			logger.Debug(ctx, "Agent metrics collection is starting")
+			timer := prometheus.NewTimer(metricsCollectorAgentStats)
+
+			checkpoint := time.Now()
+			stats, err := db.GetWorkspaceAgentStatsAndLabels(ctx, createdAfter)
+			if err != nil {
+				logger.Error(ctx, "can't get agent stats", slog.Error(err))
+				goto done
+			}
+
+			if len(stats) == 0 {
```

Comment:
This check could guard the commits below instead, which would avoid the need for goto.

Reply:
Fixed